### PR TITLE
support YAML/JSON toc metadata

### PR DESCRIPTION
**Pull request by OsmondJiang**
#3139

### CODE REVIEW

File: docs/specs/metadata.yml [hunk 1]
```diff
@@ -125,3 +125,67 @@ outputs:
...
+    }
+  build.manifest:
+---
+# yaml&json toc metadata inclusion
+inputs:
+  docfx.yml:
+  docs/json/TOC.json: |
+    {  
+      "metadata": {"a": "b"},
+      "items":[{"name":"title", "metadata": {"c": "d"},}]
```
Comment (superyyrrzz) [inline docs/specs/metadata.yml]: >, [](start = 54, length = 1)

remove `,`

File: docs/specs/metadata.yml [hunk 2]
```diff
@@ -125,3 +125,67 @@ outputs:
...
+    }
+  docs/yaml/TOC.yml: |
+    - name: toc reference
+      tocHref: ../json/TOC.json
+outputs:
+  docs/yaml/TOC.json: |
+    {  
+      "items":[{"toc_title":"toc reference", "metadata": {"a": "b"}, "children": [{"toc_title":"title", "metadata": {"c": "d"}}]}]
+    }
+  build.manifest:
```
Comment (superyyrrzz) [inline docs/specs/metadata.yml]: add some tests for metadata from config?

File: docs/specs/metadata.yml [hunk 3]
```diff
@@ -125,3 +125,96 @@ outputs:
...
+---
+# yaml&json toc leaf metadata
+inputs:
+  docfx.yml:
+  docs/json/TOC.json: |
+    [{ "name": "title", "metadata": {"a": "b"} }]
+  docs/yaml/TOC.yml: |
+    - name: title
+      metadata:
```
Comment (herohua) [inline docs/specs/metadata.yml]: >metadata [](start = 6, length = 8)

I don't think we need to support this

File: src/docfx/build/toc/parser/TableOfContentsInputItem.cs [hunk 4]
```diff
@@ -19,6 +20,8 @@ internal class TableOfContentsInputItem
 
         public string TocHref { get; set; }
 
+        public JObject Metadata { get; set; }
```
Comment (herohua) [inline src/docfx/build/toc/parser/TableOfContentsInputItem.cs]: >Metadata [](start = 23, length = 8)

per schema, this only exists in root level

Comment (OsmondJiang) [reply src/docfx/build/toc/parser/TableOfContentsInputItem.cs]: where is TOC schema? Tianqi told me that leaf metadata will be used for monikers

---
In reply to: [206754936](https://github.com/dotnet/docfx/pull/3141#discussion_r206754936) [](ancestors = 206754936)

File: docs/specs/metadata.yml [hunk 5]
```diff
@@ -125,3 +125,96 @@ outputs:
...
+      "metadata": {"a": "b"},
+      "items":[{"name":"title", "c": "d"}]
+    }
+  docs/yaml/TOC.yml: |
+    - name: toc reference
+      tocHref: ../json/TOC.json
+outputs:
+  docs/yaml/TOC.json: |
+    {  
+      "items":[{"toc_title":"toc reference", "children": [{"toc_title":"title", "c": "d"}]}]
```
Comment (superyyrrzz) [inline docs/specs/metadata.yml]: > "c": "d" [](start = 79, length = 9)

> # child's root metadata will be dropped
It's not dropped. conflict with test title?

Comment (superyyrrzz) [reply docs/specs/metadata.yml]: As discussion, ~~child~~**embeded toc**'s root metadata will be dropped

---
In reply to: [206760980](https://github.com/dotnet/docfx/pull/3141#discussion_r206760980) [](ancestors = 206760980)

Comment (OsmondJiang) [reply docs/specs/metadata.yml]: O

---
In reply to: [206762775](https://github.com/dotnet/docfx/pull/3141#discussion_r206762775) [](ancestors = 206762775,206760980)

**Pull request merged**